File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/interceptor/AbstractRedirectInterceptor.java
package com.roxiemobile.networkingapi.network.rest.interceptor;

import com.roxiemobile.androidcommons.diagnostics.Guard;
import com.roxiemobile.androidcommons.util.StringUtils;
import com.roxiemobile.networkingapi.network.http.HttpHeaders;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;

import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.internal.http.RealResponseBody;
import okio.GzipSource;
import okio.Okio;

public abstract class AbstractRedirectInterceptor implements Interceptor {

// MARK: - Methods

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request request = chain.request();

        // Execute HTTP request
        Response response = chain.proceed(request);
        if (response.isRedirect()) {

            // Handle HTTP redirects
            response = onRedirect(response);
        }

        // Done
        return response;
    }

    /**
     * TODO
     */
    public abstract Response onRedirect(@NotNull Response response) throws IOException;

    /**
     * Returns a new response that does gzip decompression on {@code response}.
     */
    protected Response decompressResponse(@NotNull Response response) throws IOException {
        Guard.notNull(response, "response is null");

        ResponseBody body = response.body();
        if (body != null && "gzip".equalsIgnoreCase(response.header(HttpHeaders.CONTENT_ENCODING))) {

            // Uncompress a response body
            GzipSource source = new GzipSource(body.source());
            Headers strippedHeaders = response.headers().newBuilder()
                    .removeAll(HttpHeaders.CONTENT_ENCODING)
                    .removeAll(HttpHeaders.CONTENT_LENGTH)
                    .build();

            // Create new HTTP response
            ResponseBody responseBody = new RealResponseBody(
                    strippedHeaders.get(HttpHeaders.CONTENT_TYPE),
                    stringToLong(strippedHeaders.get(HttpHeaders.CONTENT_LENGTH)),
                    Okio.buffer(source));

            response = response.newBuilder()
                    .headers(strippedHeaders)
                    .body(ResponseBody.create(responseBody.contentType(), responseBody.bytes()))
                    .build();
        }

        // Done
        return response;
    }

// MARK: - Private Methods

    private static long stringToLong(String s) {
        long value = -1L;

        if (StringUtils.isNotBlank(s)) {
            try {
                value = Long.parseLong(s);
            }
            catch (NumberFormatException e) {
                // Do nothing
            }
        }
        return value;
    }
}

File: sources/modules/libraries/networking-api/src/main/java/com/roxiemobile/networkingapi/network/rest/interceptor/AbstractRedirectInterceptor.kt
package com.roxiemobile.networkingapi.network.rest.interceptor

import com.roxiemobile.networkingapi.network.http.HttpHeaders
import okhttp3.Interceptor
import okhttp3.Interceptor.Chain
import okhttp3.Response
import okhttp3.ResponseBody
import okhttp3.internal.http.RealResponseBody
import okio.GzipSource
import okio.Okio
import java.io.IOException

abstract class AbstractRedirectInterceptor: Interceptor {

// MARK: - Methods

    @Throws(IOException::class)
    override fun intercept(chain: Chain): Response {
        val request = chain.request()

        // Execute HTTP request
        var response = chain.proceed(request)
        if (response.isRedirect) {

            // Handle HTTP redirects
            response = onRedirect(response)
        }

        return response
    }

    /**
     * TODO
     */
    @Throws(IOException::class)
    abstract fun onRedirect(response: Response): Response

    /**
     * Returns a new response that does gzip decompression on `response`.
     */
    @Throws(IOException::class)
    protected fun decompressResponse(response: Response): Response {
        var newResponse = response

        val responseBody = response.body()
        if (responseBody != null && "gzip".equals(response.header(HttpHeaders.CONTENT_ENCODING), ignoreCase = true)) {

            // Uncompress a response body
            GzipSource(responseBody.source()).use { gzipSource ->

                val strippedHeaders = response.headers().newBuilder()
                    .removeAll(HttpHeaders.CONTENT_ENCODING)
                    .removeAll(HttpHeaders.CONTENT_LENGTH)
                    .build()

                // Create new HTTP response
                val newResponseBody = RealResponseBody(
                    strippedHeaders[HttpHeaders.CONTENT_TYPE],
                    strippedHeaders[HttpHeaders.CONTENT_LENGTH]?.toLongOrNull() ?: -1L,
                    Okio.buffer(gzipSource),
                )

                newResponse = response.newBuilder()
                    .headers(strippedHeaders)
                    .body(ResponseBody.create(newResponseBody.contentType(), newResponseBody.bytes()))
                    .build()
            }
        }

        return newResponse
    }
}
